### PR TITLE
Handling invalid refresh token

### PR DESCRIPTION
**Pull request by mwojtaczka**

### CODE REVIEW

File: src/main/java/pl/michal/grzesiak/luxi/authentication/InvalidRefreshTokenException.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package pl.michal.grzesiak.luxi.authentication;
+
+public class InvalidRefreshTokenException extends RuntimeException {
+
+	@Override
+	public String getMessage() {
+		return "Invalid refresh token";
```
Comment (m-grzesiak) [inline src/main/java/pl/michal/grzesiak/luxi/authentication/InvalidRefreshTokenException.java:7]: should be static constant

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.mockito.ArgumentMatchers.any;
+
+@ExtendWith(MockitoExtension.class)
+class AuthenticationTest {
+
+	@Mock
+	private LuxmedApiClient luxmedApiClient;
+
+	private Authentication authentication;
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:26]: `@InjectMocks`

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+	@BeforeEach
+	void setup() {
+		authentication = AuthenticationFactory.create(luxmedApiClient);
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:30]: to remove

File: src/main/java/pl/michal/grzesiak/luxi/authentication/Authentication.java [hunk 4]
```diff
@@ -37,7 +37,8 @@ private AuthenticationException getAuthenticationException(ApiBusinessErrors api
 	public LogInResult renewSession(RefreshToken refreshToken) {
 		log.debug("Renewing user session using provided refresh token");
 		RequestParameters requestParameters = authenticationParametersFactory.createRenewSessionParameters(refreshToken);
-		return luxmedApiClient.sendPost(LOGIN, requestParameters, LogInResponse.class)
+		return luxmedApiClient.sendPost2(LOGIN, requestParameters, LogInResponse.class)
```
Comment (m-grzesiak) [inline src/main/java/pl/michal/grzesiak/luxi/authentication/Authentication.java:40]: do we still need `sendPost2`?

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	@Test
+	@DisplayName("refreshToken() should throw InvalidRefreshTokenException when error result")
+	void shouldThrowInvalidRefreshTokenException() {
+		ApiOperationResult<Object> errorResult = ApiOperationResult.ofError(new ApiBusinessErrors(List.of()));
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:36]: this should be a mock

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+	private Authentication authentication;
+
+	@BeforeEach
+	void setup() {
+		authentication = AuthenticationFactory.create(luxmedApiClient);
+	}
+
+	@Test
+	@DisplayName("refreshToken() should throw InvalidRefreshTokenException when error result")
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:34]: wrong display name.

> should throw InvalidRefreshTokenException when API Client has returned an error

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	void shouldThrowInvalidRefreshTokenException() {
+		ApiOperationResult<Object> errorResult = ApiOperationResult.ofError(new ApiBusinessErrors(List.of()));
+		Mockito.when(luxmedApiClient.sendPost2(any(), any(), any()))
+			   .thenReturn(errorResult);
+
+		Throwable thrown = catchThrowable(() -> authentication.renewSession(() -> "refresh token"));
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:40]: thrown -> obtainedException

File: src/test/java/pl/michal/grzesiak/luxi/LuxmedErrorTest.java [hunk 8]
```diff
@@ -46,6 +48,19 @@ void shouldThrowAuthenticationException() {
...
+	@Test
+	@DisplayName("should throw InvalidRefreshTokenException when refresh token is invalid")
+	void shouldThrowInvalidRefreshTokenException() {
+		RefreshToken refreshToken = () -> "invalid_refresh_token";
+
+		Throwable thrown = catchThrowable(() -> luxmed.authentication()
+													  .renewSession(refreshToken));
+
+		assertThat(thrown).isInstanceOf(InvalidRefreshTokenException.class);
+		var e = (InvalidRefreshTokenException) thrown;
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/LuxmedErrorTest.java:60]: hmm, so maybe overriding getMessage() method in the exception is not a good idea - if we pass the message through the constructor, you don't have to cast exception to specific type to retrieve the message.

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	@InjectMocks
+	private Authentication authentication;
+
+	@Test
+	@DisplayName("logIn should throw AuthenticationException when ApiClient returns an error")
+	void shouldThrowAuthenticationException() {
+		Credentials credentials = new Credentials("login", "invalid password");
+		ApiBusinessErrors.Error invalidCredentialsError =
+				new ApiBusinessErrors.Error(2, "Invalid login or password.");
+		ApiOperationResult<Object> errorResult =
```
Comment (m-grzesiak) [inline src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:35]: wrap this into separate, private method (apiOperationResult + mocking) - it is hard to read in that form

File: src/test/java/pl/michal/grzesiak/luxi/LuxmedErrorTest.java (hunk 8 above)
Comment (mwojtaczka) [reply src/test/java/pl/michal/grzesiak/luxi/LuxmedErrorTest.java]: done

File: src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java (hunk 9 above)
Comment (mwojtaczka) [reply src/test/java/pl/michal/grzesiak/luxi/authentication/AuthenticationTest.java:46]: done

File: src/test/resources/mappings/errors/invalid_token.json [hunk 10]
```diff
@@ -0,0 +1,51 @@
+{
```
Comment (m-grzesiak) [inline src/test/resources/mappings/errors/invalid_token.json:1]: Why are you going add this? We had discussion about it. We don't want to handle such cases in integration tests - you can cover it by unit test.

**Pull request merged**